File: app/src/main/java/com/jpwolfso/privdnsqt/PrivateDnsTileService.java
package com.jpwolfso.privdnsqt;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.graphics.drawable.Icon;
import android.provider.Settings;
import android.service.quicksettings.Tile;
import android.service.quicksettings.TileService;
import android.widget.Toast;

public class PrivateDnsTileService extends TileService {

    public static String DNS_MODE_OFF = "off";
    public static String DNS_MODE_AUTO = "opportunistic";
    public static String DNS_MODE_ON = "hostname";

    public void onStartListening() {
        super.onStartListening();
        String dnsmode = Settings.Global.getString(getContentResolver(), "private_dns_mode");
        Tile tile = this.getQsTile();

        if (DNS_MODE_OFF.equalsIgnoreCase(dnsmode)) {
            refreshTile(tile, Tile.STATE_INACTIVE, getString(R.string.qt_off), R.drawable.ic_dnsoff);
        } else if (DNS_MODE_AUTO.equalsIgnoreCase(dnsmode)) {
            refreshTile(tile, Tile.STATE_ACTIVE, getString(R.string.qt_auto), R.drawable.ic_dnsauto);
        } else if (DNS_MODE_ON.equalsIgnoreCase(dnsmode)) {
            String dnsprovider = Settings.Global.getString(getContentResolver(), "private_dns_specifier");
            if ((dnsprovider != null)) {
                refreshTile(tile, Tile.STATE_ACTIVE, dnsprovider, R.drawable.ic_dnson);
            } else {
                Toast.makeText(this, R.string.toast_no_permission, Toast.LENGTH_SHORT).show();
            }
        }
    }

    public void onStopListening() {
        super.onStopListening();
    }

    public void onClick() {
        super.onClick();

        final SharedPreferences togglestates = getSharedPreferences("togglestates", Context.MODE_PRIVATE);

        final boolean toggleoff = togglestates.getBoolean("toggle_off", true);
        final boolean toggleauto = togglestates.getBoolean("toggle_auto", true);
        final boolean toggleon = togglestates.getBoolean("toggle_on", true);

        String dnsprovider = Settings.Global.getString(getContentResolver(), "private_dns_specifier");

        if (hasPermission()) {
            String dnsmode = Settings.Global.getString(getContentResolver(), "private_dns_mode");
            Tile tile = this.getQsTile();
            if (DNS_MODE_OFF.equalsIgnoreCase(dnsmode)) {
                if (toggleauto) {
                    changeTileState(tile, Tile.STATE_ACTIVE, getString(R.string.qt_auto), R.drawable.ic_dnsauto, DNS_MODE_AUTO);
                } else if (toggleon) {
                    changeTileState(tile, Tile.STATE_ACTIVE, dnsprovider, R.drawable.ic_dnson, DNS_MODE_ON);
                }
            } else if (DNS_MODE_AUTO.equalsIgnoreCase(dnsmode)) {
                if (dnsprovider != null) {
                    if (toggleon) {
                        changeTileState(tile, Tile.STATE_ACTIVE, dnsprovider, R.drawable.ic_dnson, DNS_MODE_ON);
                    } else if (toggleoff) {
                        changeTileState(tile, Tile.STATE_INACTIVE, getString(R.string.qt_off), R.drawable.ic_dnsoff, DNS_MODE_OFF);
                    }
                } else {
                    if (toggleoff) {
                        changeTileState(tile, Tile.STATE_INACTIVE, getString(R.string.qt_off), R.drawable.ic_dnsoff, DNS_MODE_OFF);
                    }
                }
            } else if (DNS_MODE_ON.equalsIgnoreCase(dnsmode)) {
                if (toggleoff) {
                    changeTileState(tile, Tile.STATE_INACTIVE, getString(R.string.qt_off), R.drawable.ic_dnsoff, DNS_MODE_OFF);
                } else if (toggleauto) {
                    changeTileState(tile, Tile.STATE_ACTIVE, getString(R.string.qt_auto), R.drawable.ic_dnsauto, DNS_MODE_AUTO);
                }
            }

        } else if (!(hasPermission())) {
            Toast.makeText(this, getString(R.string.toast_no_permission), Toast.LENGTH_SHORT).show();
        }
    }

    public boolean hasPermission() {
        return checkCallingOrSelfPermission("android.permission.WRITE_SECURE_SETTINGS") != PackageManager.PERMISSION_DENIED;
    }

    public void changeTileState(Tile tile, int state, String label, int icon, String dnsmode) {
        tile.setLabel(label);
        tile.setState(state);
        tile.setIcon(Icon.createWithResource(this, icon));
        Settings.Global.putString(getContentResolver(), "private_dns_mode", dnsmode);
        tile.updateTile();
    }

    public void refreshTile(Tile tile, int state, String label, int icon) {
        tile.setState(state);
        tile.setLabel(label);
        tile.setIcon(Icon.createWithResource(this, icon));
        tile.updateTile();
    }

}

File: app/src/main/java/com/jpwolfso/privdnsqt/PrivateDnsTileService.kt
package com.jpwolfso.privdnsqt

import android.Manifest.permission.WRITE_SECURE_SETTINGS
import android.content.pm.PackageManager
import android.graphics.drawable.Icon
import android.os.Handler
import android.os.Looper
import android.provider.Settings
import android.service.quicksettings.Tile
import android.service.quicksettings.TileService
import android.widget.Toast
import androidx.annotation.StringRes
import com.jpwolfso.privdnsqt.PrivateDnsConstants.PRIVATE_DNS_MODE
import com.jpwolfso.privdnsqt.PrivateDnsConstants.PRIVATE_DNS_SPECIFIER
import com.jpwolfso.privdnsqt.SharedPreferencesHelper.Companion.SHARED_PREF_REQUIRE_UNLOCK
import com.jpwolfso.privdnsqt.SharedPreferencesHelper.Companion.SHARED_PREF_TOGGLE_AUTO
import com.jpwolfso.privdnsqt.SharedPreferencesHelper.Companion.SHARED_PREF_TOGGLE_OFF
import com.jpwolfso.privdnsqt.SharedPreferencesHelper.Companion.SHARED_PREF_TOGGLE_ON

class PrivateDnsTileService : TileService() {

    override fun onStartListening() {
        super.onStartListening()

        val dnsmode = Settings.Global.getString(contentResolver, PRIVATE_DNS_MODE)
        val tile = this.qsTile

        if (DNS_MODE_OFF.equals(dnsmode, ignoreCase = true)) {
            refreshTile(tile, Tile.STATE_INACTIVE, getString(R.string.qt_off), R.drawable.ic_dnsoff)
        } else if (DNS_MODE_AUTO.equals(dnsmode, ignoreCase = true) || dnsmode == null) {
            refreshTile(tile, Tile.STATE_ACTIVE, getString(R.string.qt_auto), R.drawable.ic_dnsauto)
        } else if (DNS_MODE_ON.equals(dnsmode, ignoreCase = true)) {
            val dnsprovider = Settings.Global.getString(contentResolver, PRIVATE_DNS_SPECIFIER)
            if (dnsprovider != null) {
                refreshTile(tile, Tile.STATE_ACTIVE, dnsprovider, R.drawable.ic_dnson)
            } else {
                showToast(R.string.toast_no_permission)
            }
        }
    }

    override fun onClick() {
        super.onClick()

        val togglestates = SharedPreferencesHelper(this)
        val isLocked = this.isSecure && this.isLocked
        val requireUnlock = togglestates.getBoolean(SHARED_PREF_REQUIRE_UNLOCK, false)

        if (!isLocked || !requireUnlock) {
            toggle()
        } else {
            unlockAndRun {
                toggle()
            }
        }
    }

    private fun toggle() {
        val togglestates = SharedPreferencesHelper(this)

        val toggleoff = togglestates.getBoolean(SHARED_PREF_TOGGLE_OFF, true)
        val toggleauto = togglestates.getBoolean(SHARED_PREF_TOGGLE_AUTO, true)
        val toggleon = togglestates.getBoolean(SHARED_PREF_TOGGLE_ON, true)

        val dnsprovider = Settings.Global.getString(contentResolver, PRIVATE_DNS_SPECIFIER)

        if (!hasPermission()) {
            showToast(R.string.toast_no_permission)
            return
        }

        val dnsmode = Settings.Global.getString(contentResolver, PRIVATE_DNS_MODE)
        val tile = this.qsTile
        if (DNS_MODE_OFF.equals(dnsmode, ignoreCase = true)) {
            if (toggleauto) {
                changeTileState(tile, Tile.STATE_ACTIVE, getString(R.string.qt_auto), R.drawable.ic_dnsauto, DNS_MODE_AUTO)
            } else if (toggleon) {
                changeTileState(tile, Tile.STATE_ACTIVE, dnsprovider, R.drawable.ic_dnson, DNS_MODE_ON)
            }
        } else if (DNS_MODE_AUTO.equals(dnsmode, ignoreCase = true) || dnsmode == null) {
            if (dnsprovider != null) {
                if (toggleon) {
                    changeTileState(tile, Tile.STATE_ACTIVE, dnsprovider, R.drawable.ic_dnson, DNS_MODE_ON)
                } else if (toggleoff) {
                    changeTileState(tile, Tile.STATE_INACTIVE, getString(R.string.qt_off), R.drawable.ic_dnsoff, DNS_MODE_OFF)
                }
            } else {
                if (toggleoff) {
                    changeTileState(tile, Tile.STATE_INACTIVE, getString(R.string.qt_off), R.drawable.ic_dnsoff, DNS_MODE_OFF)
                }
            }
        } else if (DNS_MODE_ON.equals(dnsmode, ignoreCase = true)) {
            if (toggleoff) {
                changeTileState(tile, Tile.STATE_INACTIVE, getString(R.string.qt_off), R.drawable.ic_dnsoff, DNS_MODE_OFF)
            } else if (toggleauto) {
                changeTileState(tile, Tile.STATE_ACTIVE, getString(R.string.qt_auto), R.drawable.ic_dnsauto, DNS_MODE_AUTO)
            }
        }
    }

    private fun hasPermission(): Boolean {
        return checkCallingOrSelfPermission(WRITE_SECURE_SETTINGS) != PackageManager.PERMISSION_DENIED
    }

    private fun changeTileState(tile: Tile, state: Int, label: String?, icon: Int, dnsmode: String?) {
        tile.label = label
        tile.state = state
        tile.icon = Icon.createWithResource(this, icon)
        Settings.Global.putString(contentResolver, PRIVATE_DNS_MODE, dnsmode)
        tile.updateTile()
    }

    private fun refreshTile(tile: Tile, state: Int, label: String?, icon: Int) {
        tile.state = state
        tile.label = label
        tile.icon = Icon.createWithResource(this, icon)
        tile.updateTile()
    }

    private fun showToast(@StringRes resId: Int) {
        Handler(Looper.getMainLooper()).post {
            Toast.makeText(applicationContext, resId, Toast.LENGTH_SHORT).show()
        }
    }

    companion object {
        const val DNS_MODE_OFF = "off"
        const val DNS_MODE_AUTO = "opportunistic"
        const val DNS_MODE_ON = "hostname"
    }
}